### PR TITLE
Feature EPL-1357: Connector must retry from last failed step

### PR DESCRIPTION
**Pull request by sebastianbarrozo**

### CODE REVIEW

File: libs/com.etendorx.lib.kafka/src/main/java/com/etendorx/lib/kafka/KafkaMessageUtil.java
Comment (etendobot) [inline libs/com.etendorx.lib.kafka/src/main/java/com/etendorx/lib/kafka/KafkaMessageUtil.java:1]: # GPT Review for KafkaMessageUtil.java
This file has no corrections to be made. Well done!

File: libs/com.etendorx.lib.kafka/src/main/java/com/etendorx/lib/kafka/topology/AsyncProcessTopology.java
Comment (etendobot) [inline libs/com.etendorx.lib.kafka/src/main/java/com/etendorx/lib/kafka/topology/AsyncProcessTopology.java:1]: # GPT Review for AsyncProcessTopology.java
1. **Parameterize the `buildTopology` method**: Instead of creating a new `StreamsBuilder` object inside the method, consider passing it as a parameter. This change allows for better flexibility and reusability of the method, as it can work with different `StreamsBuilder` instances.

2. **Update method signature**: Since the method is now void and requires a `StreamsBuilder` parameter, make sure to update the method signature to reflect these changes. This will help other developers understand the purpose and requirements of the method more clearly.

3. **Remove redundant object creation**: With the new approach of passing `StreamsBuilder` as a parameter, the creation of `StreamsBuilder` object inside the method becomes redundant. Removing this unnecessary object creation will improve the code's efficiency and readability.

4. **Consider adding comments**: Depending on the complexity of the logic inside the method, consider adding comments to explain the purpose of different parts of the code. This will make it easier for other developers to understand the functionality without diving deep into the implementation details.

5. **Ensure consistency in naming**: Make sure that variable names are consistent and follow a clear naming convention. For example, consider using `streamsBuilder` consistently throughout the method to refer to the `StreamsBuilder` parameter passed to the method. This consistency improves code readability.

File: modules_core/com.etendorx.asyncprocess/src/main/java/com/etendorx/asyncprocess/config/LatestLogsConfiguration.java
Comment (etendobot) [inline modules_core/com.etendorx.asyncprocess/src/main/java/com/etendorx/asyncprocess/config/LatestLogsConfiguration.java:1]: # GPT Review for LatestLogsConfiguration.java
1. **Use Descriptive Variable Names**: Consider using more descriptive variable names to improve code readability. For example, instead of `list`, you could use `logsQueue` or `logsList` to make the purpose of the variable clearer.

2. **Add Comments for Complex Logic**: While the code is well-structured, adding comments to explain the logic behind certain operations, especially in the `updateList` method, can help other developers understand the code more easily.

3. **Error Handling**: Ensure that error handling is in place, especially when dealing with operations like adding or removing elements from a list. Consider adding try-catch blocks or validating inputs to prevent potential runtime errors.

4. **Optimization**: The `updateList` method could be optimized further by avoiding unnecessary operations. For example, instead of calling `value.getExecutions().clear()` unconditionally, you could check if the list needs to be cleared based on certain conditions.

5. **Testing**: It's important to thoroughly test the functionality of the `updateList` method, especially edge cases like when the list is empty or when the maximum size is reached.

6. **Consistency in Naming**: Ensure consistency in naming conventions throughout the codebase. For instance, the `lastRecords` method could be renamed to `aggregateLatestLogs` for better consistency.

7. **Encapsulation**: Consider encapsulating the logic related to Kafka Streams API operations into separate methods for better modularity and maintainability.

8. **Documentation**: Document the public methods with Javadoc comments to describe their purpose, parameters, and return values clearly.

By considering these suggestions, you can enhance the readability, maintainability, and robustness of the codebase.

File: modules_core/com.etendorx.asyncprocess/src/main/java/com/etendorx/asyncprocess/config/StreamConfiguration.java
Comment (etendobot) [inline modules_core/com.etendorx.asyncprocess/src/main/java/com/etendorx/asyncprocess/config/StreamConfiguration.java:1]: # GPT Review for StreamConfiguration.java
1. **Consistent Naming Convention**: It's great that you have added comments to describe the purpose of each method and variable. To further improve readability, consider following a consistent naming convention throughout the class. For example, you can use camelCase for method names and variables to adhere to Java conventions.

2. **Use of StreamsBuilder**: Utilizing `StreamsBuilder` to build the topology is a good practice. However, ensure that you are consistent in using it. In the `kafkaStreams` method, you are creating a `StreamsBuilder` instance, but in the `AsyncProcessTopology.buildTopology` method, you are not using the provided `StreamsBuilder`. Make sure to use the same `StreamsBuilder` instance consistently for building the topology.

3. **Error Handling**: It's important to consider error handling in your Kafka Streams application. Ensure that you have mechanisms in place to handle exceptions that may occur during the Kafka Streams processing. This will make your application more robust and reliable.

4. **Configuration Properties**: When defining configuration properties, consider externalizing them to a configuration file (e.g., application.properties) instead of hardcoding them in the code. This allows for easier configuration management without the need to modify the code.

5. **Code Duplication**: In the `kafkaStreamsConfiguration` method, the properties related to Kafka Streams configuration are being set. If these properties are used in multiple places, consider extracting them into a separate method to avoid code duplication and improve maintainability.

6. **Documentation**: While you have added method-level comments, consider adding class-level comments to provide an overview of the purpose of the `StreamConfiguration` class.

7. **Testing**: Don't forget to thoroughly test your Kafka Streams application to ensure that it behaves as expected under different scenarios, including edge cases and failure conditions.

By considering these suggestions, you can enhance the readability, maintainability, and robustness of your Kafka Streams configuration code.

File: modules_core/com.etendorx.asyncprocess/src/main/java/com/etendorx/asyncprocess/controller/AsyncProcessController.java
Comment (etendobot) [inline modules_core/com.etendorx.asyncprocess/src/main/java/com/etendorx/asyncprocess/controller/AsyncProcessController.java:1]: # GPT Review for AsyncProcessController.java
In the provided git diff, a new method `getLatestAsyncProcess` has been added to the `AsyncProcessController` class. Here are some recommendations to improve this code:

1. **Method Naming**: The method name `getLatestAsyncProcess` implies that it will return a single `AsyncProcess`, but it actually returns a list of `AsyncProcess`. Consider renaming the method to `getLatestAsyncProcesses` to better reflect that it returns multiple processes.

2. **Documentation**: Add comments or JavaDoc to describe the purpose of the `getLatestAsyncProcesses` method, its parameters (if any), and what it returns. This will help other developers understand the functionality without diving into the implementation details.

3. **Error Handling**: Consider adding error handling mechanisms to deal with potential exceptions that may occur during the execution of `getLatestAsyncProcesses`. This could include handling cases where the service call fails or returns null.

4. **HTTP Method**: Ensure that using `@GetMapping` with a void method like `getLatestAsyncProcesses` is appropriate. If this method doesn't actually handle HTTP GET requests, consider using a different annotation or refactor the method accordingly.

5. **Consistency**: Make sure that the naming conventions and coding style are consistent with the rest of the class and the project to maintain readability and maintainability.

6. **Testing**: Don't forget to add unit tests for the new method to ensure its correctness and functionality. This will help in catching any regressions in the future.

By considering these recommendations, you can enhance the readability, maintainability, and robustness of the `AsyncProcessController` class.

File: modules_core/com.etendorx.asyncprocess/src/main/java/com/etendorx/asyncprocess/serdes/DequeDeserializer.java
Comment (etendobot) [inline modules_core/com.etendorx.asyncprocess/src/main/java/com/etendorx/asyncprocess/serdes/DequeDeserializer.java:1]: # GPT Review for DequeDeserializer.java
1. **Documentation**: The code includes good documentation with comments explaining the purpose of each method and class. This is a good practice for code maintainability and readability.

2. **Error Handling**: The `deserialize` method includes error handling for `IOException` by throwing a `SerdesException`. This is a good practice to handle exceptions and provide meaningful error messages.

3. **Resource Management**: The `DataInputStream` is created within a try-with-resources block, ensuring that it is closed properly after its use. This is a good practice to manage system resources efficiently.

4. **SneakyThrows**: The `@SneakyThrows` annotation is used in the `deserialize` method. While this can reduce boilerplate code, it may hide exceptions and make debugging more challenging. Consider handling exceptions explicitly for better transparency.

5. **Variable Naming**: The variable names are clear and descriptive, enhancing code readability. However, consider using more specific names instead of generic ones like `s` and `T` to improve clarity.

6. **Performance**: The deserialization process involves reading integers and byte arrays from a `DataInputStream`. Consider optimizing the deserialization process if performance is critical, as reading byte by byte can be inefficient for large data sets.

7. **Testing**: Ensure to thoroughly test the deserialization logic with different types of data to validate its correctness and robustness.

8. **Future Improvements**: Depending on the requirements, you may consider adding additional features such as custom serialization logic, handling different data formats, or supporting more advanced deserialization scenarios.

Overall, the code is well-structured and follows good practices. Consider the mentioned suggestions to further enhance the code quality and maintainability.

File: modules_core/com.etendorx.asyncprocess/src/main/java/com/etendorx/asyncprocess/serdes/DequeSerializer.java
Comment (etendobot) [inline modules_core/com.etendorx.asyncprocess/src/main/java/com/etendorx/asyncprocess/serdes/DequeSerializer.java:1]: # GPT Review for DequeSerializer.java
1. **Documentation**: The code includes good documentation with comments explaining the purpose of each method and the class itself. This is a good practice for code maintainability and readability.

2. **Exception Handling**: The `serialize` method includes a try-catch block to handle `IOException`. This is important for robustness, but consider logging the exception details for better debugging information.

3. **Resource Management**: The `DataOutputStream` should be closed in a `finally` block to ensure proper resource management, even if an exception is thrown during serialization.

4. **Serialization Logic**: The serialization logic seems straightforward and efficient. However, consider validating input parameters like `topic` and `queue` for null values to prevent potential NullPointerExceptions.

5. **Naming**: The variable names are clear and descriptive. Consider following a consistent naming convention throughout the codebase for better readability.

6. **Unused Methods**: The `configure` and `close` methods are not used in this implementation. Consider removing them if they are not necessary, to keep the codebase clean and concise.

7. **License Information**: Ensure that the license information provided in the file is accurate and up to date.

8. **Testing**: It's important to thoroughly test the `DequeSerializer` class to ensure that it works as expected in different scenarios, including edge cases and error conditions.

Overall, the code looks well-structured and follows good practices. Consider the suggestions provided to further enhance the code quality and maintainability.

File: modules_core/com.etendorx.asyncprocess/src/main/java/com/etendorx/asyncprocess/serdes/PriorityQueueSerde.java
Comment (etendobot) [inline modules_core/com.etendorx.asyncprocess/src/main/java/com/etendorx/asyncprocess/serdes/PriorityQueueSerde.java:1]: # GPT Review for PriorityQueueSerde.java
1. **Use of Generics**: Utilizing generics in the `PriorityQueueSerde` class is a good practice as it allows for flexibility in the type of objects stored in the Deque. However, ensure that the generic type `<T>` is appropriately constrained to avoid potential issues with type safety.

2. **Constructor Parameter Naming**: Consider using a more descriptive name than `avroSerde` for the parameter in the constructor of `PriorityQueueSerde`. This can improve code readability and make it clearer what the parameter represents.

3. **Error Handling**: It's important to handle any potential exceptions that may occur during serialization or deserialization. Adding try-catch blocks or throwing exceptions can make the code more robust.

4. **Documentation**: The existing comments provide a good overview of the class and its methods. Consider adding more detailed comments to explain the purpose of specific methods or any complex logic to make it easier for other developers to understand the code.

5. **Resource Management**: Ensure that resources are properly managed, especially in the `close()` method. Make sure to release any resources held by the serializer and deserializer to prevent memory leaks.

6. **Consistency in Naming**: Ensure consistency in naming conventions throughout the class. For example, the parameter `avroSerde` could be renamed to follow the camelCase convention consistently.

7. **Testing**: Don't forget to thoroughly test the functionality of the `PriorityQueueSerde` class, especially the serialization and deserialization processes, to ensure they work as expected.

8. **Dependency Injection**: Consider using dependency injection to provide the inner Serde object instead of creating it within the constructor. This can improve the flexibility and testability of the class.

File: modules_core/com.etendorx.asyncprocess/src/main/java/com/etendorx/asyncprocess/service/AsyncProcessService.java
Comment (etendobot) [inline modules_core/com.etendorx.asyncprocess/src/main/java/com/etendorx/asyncprocess/service/AsyncProcessService.java:1]: # GPT Review for AsyncProcessService.java
1. **Code Documentation**:
   - It's great to see the addition of comments explaining the purpose of the class and methods. Consider adding more detailed comments to describe the functionality of each method and any important parameters or return values.

2. **Variable Naming**:
   - The variable names `kafkaStreams` and `hostInfo` are clear and descriptive. Good job on that! 
   - Consider using more descriptive names for methods like `getStore()` and `getLatestAsyncProcesses()` to better convey their purpose.

3. **Error Handling**:
   - It's important to handle potential null values properly. In the `getLatestAsyncProcesses()` method, ensure that appropriate error handling is in place if `store` or `records` are null.

4. **Code Optimization**:
   - In the `getLatestAsyncProcesses()` method, consider using a more efficient way to handle the null check. Instead of creating a new `ArrayList` if `records` is null, you could directly return `Collections.emptyList()`.

5. **Consistency**:
   - Ensure consistency in the style of comments and method documentation throughout the class for better readability and maintainability.

6. **Resource Management**:
   - Since `kafkaStreams` and `hostInfo` are dependencies injected via constructor, ensure proper resource management and potential cleanup if needed, especially in more complex scenarios.

7. **Testing**:
   - Don't forget to thoroughly test the functionality of these methods, especially when interacting with external systems like Kafka. Consider writing unit tests to cover different scenarios.

By considering these suggestions, you can enhance the readability, maintainability, and robustness of the `AsyncProcessService` class.

**Review by giuliodoriEtendo — APPROVED** [review]

File: modules_core/com.etendorx.asyncprocess/src/main/java/com/etendorx/asyncprocess/config/LatestLogsConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+   * This method is used to update the list of logs for an asynchronous process.
+   * It removes any existing logs with the same ID as the new log and adds the new log to the front of the list.
+   * If the list size exceeds 100, it removes the oldest log from the list.
+   *
+   * @param value The new log for the asynchronous process.
+   * @param list The current list of logs for the asynchronous process.
+   * @return Deque<AsyncProcess> The updated list of logs for the asynchronous process.
+   */
+  private static Deque<AsyncProcess> updateList(AsyncProcess value, Deque<AsyncProcess> list) {
+    if (value.getId() == null || value.getState().equals(AsyncProcessState.ACCEPTED)) {
```
Comment (Matias-Bernal) [inline modules_core/com.etendorx.asyncprocess/src/main/java/com/etendorx/asyncprocess/config/LatestLogsConfiguration.java:80]: StringUtils

File: modules_core/com.etendorx.asyncprocess/src/main/java/com/etendorx/asyncprocess/config/LatestLogsConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+      return list;
+    }
+    list.removeIf(existingValue -> existingValue.getId().equals(value.getId()));
```
Comment (Matias-Bernal) [inline modules_core/com.etendorx.asyncprocess/src/main/java/com/etendorx/asyncprocess/config/LatestLogsConfiguration.java:83]: StringUtils

**Review by Matias-Bernal — APPROVED** [review]
